### ISSUE TITLE
Github commands package crashes all bot features if github keey is not configured

### ISSUE THREAD

**Issue by ankitsmt211** [Together-Java/TJ-Bot#1020]
**Describe the bug**

Ideally when setting up bot, it should work with copy paste of `config.template` and bot token only. But atm, without entering githubAPI key, none of the other features work.

The root cause seems to be contructors of both `GithubCommand.java` and `GithubReference.java`, both calls two methods `updateCache()` and `acquireRespositories()`, but since we havent configured key yet this ends up crashing all other features.

**Expected behavior**

Other features except `/github-search` and `GithubReference` should work without githubAPI key.

**To Reproduce**

1. Clone Bot.
2. Copy `config.json.template` and create a new file inside `application` folder named `config.json`.
3. Add discord bot token.
4. When bot is live, use `/ping` or any other feature.

**Additional context**

constructor of `GithubCommand`

![image](https://github.com/Together-Java/TJ-Bot/assets/61616007/f5e8bd91-710e-4d99-8c2a-efb82687540d)

constructor of `GithubReference`
![image](https://github.com/Together-Java/TJ-Bot/assets/61616007/88a94627-cc75-4d87-ac62-ec4fd09448bd)

**Comment by ankitsmt211** [comment]
@marko-radosavljevic The issue i already mentioned if you wanna have a look.

**Comment by marko-radosavljevic** [comment]
That's indeed a problem, even if we disable this feature, it would still crash. Because object creation happens before we filter disabled features.

The most obvious solution seems moving these out of the constructor. So, for the first class, the only user of that method `updateCache()` is `onAutoComplete()`, and there it is already called. So maybe update cache at the start of the method (in a way that would run it the first time it is called as well).

In the similar fashion for the other class, you can do that invocation much latter, when `findIssue()` needs it. Since there are two different entries, check if it's already populated.

And then the feature can be disabled, for people that do not care about that feature or don't have a key.

Also, when if there is a bad key, or default "<insert your key here>", it should throw a friendly exception that tells the user exactly where is the problem. "The github key ({}) used in the config is invalid, these features .. will not work."

Because another issue is that our config is getting more complicated, and requires a ton of different stuff. If you don't have the key, are you supposed to leave it empty or write null? You do that, and you get NPE, whole thing crashes, alright. They told me I can use dummy data, but when I use `foo` it crashes as invalid URL `IllegalArgumentException: Failed to parse webhook URL`. Alright so webhook needs dummy data, but not too dummy, it needs to be an actual url, that doesn't resolve... etc

So the purpose of our template was to make it easy to start the bot, it's already populated with everything you need to get started, so the bot doesn't crash. It should be simple, and consistent.

**Comment by ankitsmt211** [comment]
moving methods out of constructor only works for `GithubReference` but not with `GithubCommand`, because incase of reference it's an interaction whereas in case of `GithubCommand`, it should auto-suggest issues on typing `/github-search` only. 

https://javadoc.io/doc/net.dv8tion/JDA/latest/net/dv8tion/jda/api/events/interaction/GenericAutoCompleteInteractionEvent.html

This event could help with updating cache on typing, but i havent thought about specifics to implement it. Shouldn't this be part of `BotCore.java`? Any ideas to go about this?

**Comment by ankitsmt211** [comment]
@marko-radosavljevic i found a way to solve this without moving methods out of constructor since if you're not configuring those keys there's a solid chance you are not going to use it anyway, 

  ```java
  try{
            GitHubReference githubReference = new GitHubReference(config);
            GitHubCommand gitHubCommand = new GitHubCommand(githubReference);
            features.add(githubReference);
            features.add(gitHubCommand);
        }
        catch (Exception exception){
            LoggerFactory.getLogger(Features.class).error("github key is not configured, feature will not work",exception);
        }

        ```
        
     This ofc is a dirty snippet sitting in `Features.java` atm, Any suggestions?

**Comment by Tais993** [comment]
This goes back to the fact that none of the features have a way to be disabled universally. 

A global system should be added so optional features have to be enabled in the config manually before actually being enabled / or they disable automatically when lacking the valid information or when empty.

We talked before about this, and I don't remember what was decided in the end.

**Comment by ankitsmt211** [comment]
@Tais993 i think you're referring to this PR which is already merged. 
https://github.com/Together-Java/TJ-Bot/pull/886

**Comment by marko-radosavljevic** [comment]
Yeah, the issue is that the feature object creation happens before adding them to the blacklist. And calls to APIs that are not specified in config happen during construction. So blacklisted or not, this crashes. 

```java
        features.add(new BookmarksCommand(bookmarksSystem));
        features.add(new ChatGptCommand(chatGptService));
        features.add(new JShellCommand(jshellEval));


        FeatureBlacklist<Class<?>> blacklist = blacklistConfig.normal();
        return features.stream().filter(f -> blacklist.isEnabled(f.getClass())).toList();
    }
```
https://github.com/Together-Java/TJ-Bot/blob/214882178d598408a496358a8452122d0a8aa296/application/src/main/java/org/togetherjava/tjbot/features/Features.java#L164-L170

Probably the simplest and most graceful solution for now is just not to call any APIs specified in config during construction, but when it's actually used/needed, like every other feature. If that is not possible for some technical reason, you can do your try catch, but in classes where the API is actually called. With some nice, and user-friendly explanation, that the API key is not provided in the config, and thus the feature is disabled.

And if this problem happens again, we can think of redesigning disabling features and config.
